### PR TITLE
feat(leaderelection): impl leader election for HA Deployment

### LINKED ISSUE #720: HA Mode for Deployment using Leader Election
**Is your feature request related to a problem? Please describe.**
I'm not so sure whether it has been discussed before or not but dropping the idea here so we can discuss further it.

Assume that one of the nodes that contain your _descheduler pod_ has crashed. If we keep auto-scaling the nodes by joining/removing to cluster, so descheduler couldn't do its job properly for a while until the broken node gets fixed. This is where I came up with this idea.

**Describe the solution you'd like**

* Implement [leaderelection](https://pkg.go.dev/k8s.io/client-go/tools/leaderelection) client
  * One of the pods should be leader
  * It should set `leases.coordination.k8s.io`
* Ability to set more than `1` pods for deployment

**Describe alternatives you've considered**
Are we really want to run in HA mode? Would it be over-engineered?

**What version of descheduler are you using?**

descheduler version: `0.23.1`

**Additional context**
`-`

### PR DESCRIPTION
**Pull request by Dentrax**
Fixes #720

**TODO:**
- [x] update README
- [X] pass namespace dynamically (used `BindLeaderElectionFlags`)
- [X] pass durations dynamically from cmdOptions (used LeaderElectionFlag set)
- [X] better id handle: POD-NAME_UUID (used `os.Hostname` + UUID)
- [X] add e2e test
- [X] update Helm Chart (Pod Static 1)
  - [X] add leaderElection config
  - [X] add podAntiAffinity example
  - [X] enforce replica should not greater than 1 if leaderElection is not set

Signed-off-by: Furkan <furkan.turkal@trendyol.com>
Co-authored-by: Emin <emin.aktas@trendyol.com>
Co-authored-by: Yasin <yasintaha.erol@trendyol.com>

### CODE REVIEW

File: charts/descheduler/templates/deployment.yaml [hunk 1]
```diff
@@ -6,7 +6,14 @@ metadata:
   labels:
     {{- include "descheduler.labels" . | nindent 4 }}
 spec:
+  {{- if gt .Values.replicas 1.0}}
+  {{- if and .Values.leaderElection .Values.leaderElection.enabled }}
```
Comment (damemi) [inline charts/descheduler/templates/deployment.yaml:10]: it could also be helpful to check that users aren't doing the inverse, ie setting >1 replicas without leaderelection

File: pkg/descheduler/descheduler.go [hunk 2]
```diff
@@ -75,7 +76,21 @@ func Run(ctx context.Context, rs *options.DeschedulerServer) error {
 		<-ctx.Done()
 		close(stopChannel)
 	}()
-	return RunDeschedulerStrategies(ctx, rs, deschedulerPolicy, evictionPolicyGroupVersion, stopChannel)
+
+	runFn := func() error {
+		return RunDeschedulerStrategies(ctx, rs, deschedulerPolicy, evictionPolicyGroupVersion, stopChannel)
+	}
```
Comment (damemi) [inline pkg/descheduler/descheduler.go:82]: It feels a bit odd to be wrapping the function like this, is there a different way we could pass it to `leader.NewLeaderElection` with all of the parameters intact?

File: pkg/descheduler/descheduler.go [hunk 3]
```diff
@@ -75,7 +76,21 @@ func Run(ctx context.Context, rs *options.DeschedulerServer) error {
...
+
+	// if deployed as a Deployment,
+	if rs.LeaderElection.LeaderElect {
+		if rs.DeschedulingInterval.Seconds() > 0 {
+			leader.NewLeaderElection(runFn, rsclient, rs.LeaderElection, ctx)
+			return nil
+		}
+		klog.V(1).InfoS("Warning: You set LeaderElection to True but server does not use Deployment kind.")
```
Comment (damemi) [inline pkg/descheduler/descheduler.go:90]: technically `DeschedulingInterval` can be set on cronjobs too, so it's not purely indicative of the kind. Maybe reword this error to not imply that (for example a user would be confused if they are using Deployment kind but forgot to set DeschedulingInterval)

**Review by damemi — COMMENTED** [review]
/ok-to-test
I had mostly nits on docs and logging, but the overall idea looks pretty straightforward. Thanks for the contribution @Dentrax !

File: README.md [hunk 4]
```diff
@@ -89,6 +91,9 @@ kubectl create -f kubernetes/base/configmap.yaml
 kubectl create -f kubernetes/deployment/deployment.yaml
 ```
 
+Deployment starts with 1 replica by default. If you want use more than 1 replica, please refer to 
+[High Availability](#high-availability) mode since we don't want to run descheduler pods simultaneously.
```
[the remark this answers is not in the document]
Comment (Dentrax) [reply README.md]: Makes sense! Rephrased and moved to `HA` section.

File: README.md [hunk 5]
```diff
@@ -761,6 +766,22 @@ Setting `--v=4` or greater on the Descheduler will log all reasons why any pod i
...
+In High Availability mode, Descheduler starts [leader election](https://github.com/kubernetes/client-go/tree/master/tools/leaderelection) process in Kubernetes. You can activate HA mode
+if you choose to deploy your application as Deployment.
+
+### Configure HA Mode
+
+You can install descheduler in HA using Helm Chart.
+
+To get best results from HA mode some additional configurations might require:
+* Configure [podAntiAffinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node) in the Helm Chart
+* Set replica count to at least 3
```
[the remark this answers is not in the document]
Comment (Dentrax) [reply README.md]: Clarified!

File: charts/descheduler/templates/deployment.yaml (hunk 1 above)
Comment (Dentrax) [reply charts/descheduler/templates/deployment.yaml:10]: Actually, the following condition already checks it, isn't it?

```
{{- if gt .Values.replicas 1.0}}
{{- if and .Values.leaderElection .Values.leaderElection.enabled }}
```

it means: _setting >1 replicas without leaderelection_

File: pkg/descheduler/descheduler.go (hunk 2 above)
Comment (Dentrax) [reply pkg/descheduler/descheduler.go:82]: Had same feeling. But couldn't find a better way to pass `RunDeschedulerStrategies()` func to `leader.NewLeaderElection`, any ideas and advices here?

**Comment by eminaktas** [comment]
There is an error happening [here](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_descheduler/722/pull-descheduler-verify-master/1495740448386322432#1:build-log.txt%3A163). @Dentrax and I tried to add it manually but it didn't work. Can you help us with this?

Thanks

cc @damemi @ingvagabund @Dentrax

File: pkg/descheduler/descheduler.go (hunk 2 above)
Comment (ingvagabund) [reply pkg/descheduler/descheduler.go:82]: Does `NewLeaderElection` need to be under a new directory? The function might go under `pkg/descheduler/descheduler.go` or a new `pkg/descheduler/leaderelection.go`. I do not see the `NewLeaderElection` function to be used anywhere but here so it could be just inlined:
```
if rs.LeaderElection.LeaderElect {
		if rs.DeschedulingInterval.Seconds() > 0 {
			...
			leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
					Lock:            lock,
					ReleaseOnCancel: true,
					LeaseDuration:   LeaderElectionConfig.LeaseDuration.Duration,
					RenewDeadline:   LeaderElectionConfig.RenewDeadline.Duration,
					RetryPeriod:     LeaderElectionConfig.RetryPeriod.Duration,
					Callbacks: leaderelection.LeaderCallbacks{
						OnStartedLeading: func(ctx context.Context) {
							klog.V(1).InfoS("Started leading")
							err := RunDeschedulerStrategies(ctx, rs, deschedulerPolicy, evictionPolicyGroupVersion, stopChannel)
							if err != nil {
								klog.Error(err)
							}
						},
			...
			return nil
		}
		...
}
```

File: pkg/descheduler/descheduler.go [hunk 6]
```diff
@@ -75,7 +76,21 @@ func Run(ctx context.Context, rs *options.DeschedulerServer) error {
...
+		return RunDeschedulerStrategies(ctx, rs, deschedulerPolicy, evictionPolicyGroupVersion, stopChannel)
+	}
+
+	// if deployed as a Deployment,
+	if rs.LeaderElection.LeaderElect {
+		if rs.DeschedulingInterval.Seconds() > 0 {
+			leader.NewLeaderElection(runFn, rsclient, rs.LeaderElection, ctx)
+			return nil
+		}
+		klog.V(1).InfoS("Warning: DeschedulingInterval must be greater than 0 to use LeaderElection")
```
Comment (ingvagabund) [inline pkg/descheduler/descheduler.go:90]: Would you mind returning an error here? If the leader election is enabled but can not proceed, we need to error and tell the user to set the descheduling interval.

**Comment by ingvagabund** [comment]
> T[here](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_descheduler/722/pull-descheduler-verify-master/1495740448386322432#1:build-log.txt%3A163) is an error happening here. @Dentrax and I tried to add it manually but it didn't work. Can you help us with this?

@eminaktas @Dentrax did you try running `./hack/update-generated-conversions.sh`?

**Comment by eminaktas** [comment]
> > There is an error happening [here](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_descheduler/722/pull-descheduler-verify-master/1495740448386322432#1:build-log.txt%3A163). @Dentrax and I tried to add it manually but it didn't work. Can you help us with this?
> 
> @eminaktas @Dentrax did you try running `./hack/update-generated-conversions.sh`?

Yes, we did but it returns an irrational output in the cloud.

```go
+	// WARNING: in.LeaderElection requires manual conversion: does not exist in peer-type
 	out.Logging = in.Logging
 	return nil
 }
-
-// Convert_componentconfig_DeschedulerConfiguration_To_v1alpha1_DeschedulerConfiguration is an autogenerated conversion function.
-func Convert_componentconfig_DeschedulerConfiguration_To_v1alpha1_DeschedulerConfiguration(in *componentconfig.DeschedulerConfiguration, out *DeschedulerConfiguration, s conversion.Scope) error {
-	return autoConvert_componentconfig_DeschedulerConfiguration_To_v1alpha1_DeschedulerConfiguration(in, out, s)
-}
```

When I tried it in my local, It verifies but gives me this output with the current state of this PR:

```bash
➜  descheduler git:(feature/leaderelection) ./hack/update-generated-conversions.sh 
E0222 16:22:22.731871   49961 conversion.go:755] Warning: could not find nor generate a final Conversion function for sigs.k8s.io/descheduler/pkg/apis/componentconfig.DeschedulerConfiguration -> sigs.k8s.io/descheduler/pkg/apis/componentconfig/v1alpha1.DeschedulerConfiguration
E0222 16:22:22.732092   49961 conversion.go:756]   the following fields need manual conversion:
E0222 16:22:22.732100   49961 conversion.go:758]       - LeaderElection
➜  descheduler git:(feature/leaderelection) ./hack/verify-conversions.sh 
E0222 16:23:14.442056   50171 conversion.go:755] Warning: could not find nor generate a final Conversion function for sigs.k8s.io/descheduler/pkg/apis/componentconfig.DeschedulerConfiguration -> ./pkg/apis/componentconfig/v1alpha1.DeschedulerConfiguration
E0222 16:23:14.442272   50171 conversion.go:756]   the following fields need manual conversion:
E0222 16:23:14.442280   50171 conversion.go:758]       - LeaderElection
Generated conversions verified.
```

**Comment by damemi** [comment]
@eminaktas you might need to add the new [`LeaderElection` field](https://github.com/kubernetes-sigs/descheduler/pull/722/files#diff-8cb3a832fd89345b8caf52788b477463aaf7cd78c7a947d55eef89fdc1d5e12fR57) to `pkg/apis/componentconfig/v1alpha1/types.go` as well, so that it can generate the auto-conversion between versions

File: charts/descheduler/templates/deployment.yaml (hunk 1 above)
Comment (damemi) [reply charts/descheduler/templates/deployment.yaml:10]: Yup, at least if I'm reading it right, currently this code will fail if `replicas > 1` and `leaderElection != nil`, but we want what you just posted

File: pkg/descheduler/descheduler.go (hunk 2 above)
Comment (Dentrax) [reply pkg/descheduler/descheduler.go]: We removed the `leader` package and move it into `leaderelection.go` file. I think `NewLeaderElection` function makes things cleaner in a separated file, and it would a bit verbose to put entire logic in the `descheduler.go` file. But I agree with you, we don't use _that_ function again anywhere. 🤔

Comment (damemi) [reply pkg/descheduler/descheduler.go]: Instead of nesting the `interval` check inside the `leaderElect` check, you could break both of those into an error check above. That might open up an option to make this flow better. Something like:
```go
if rs.LeaderElection.LeaderElect && rs.DeschedulingInterval.Seconds() == 0 {
  return fmt.Errorf("LeaderElection must be used with deschedulingInterval...")
}
if rs.LeaderElection.LeaderElect {
  // start leader election call
} else {
  // start normal run call
}
```

something like that. it's a little redundant but it makes the code clearer and removes the deep nesting

**Comment by eminaktas** [comment]
> @eminaktas you might need to add the new [`LeaderElection` field](https://github.com/kubernetes-sigs/descheduler/pull/722/files#diff-8cb3a832fd89345b8caf52788b477463aaf7cd78c7a947d55eef89fdc1d5e12fR57) to `pkg/apis/componentconfig/v1alpha1/types.go` as well, so that it can generate the auto-conversion between versions

I had to change a few things in the code, now it is fine. Thanks for the advice.

File: charts/descheduler/templates/deployment.yaml (hunk 1 above)
Comment (Dentrax) [reply charts/descheduler/templates/deployment.yaml]: We tested the following scenarios:

1. Works as is
```yaml
replicas: 1
leaderElection: {}
```

2. Got: `You must set leaderElection to use more than 1 replica`
```yaml
replicas: 2
leaderElection: {}
```

3. Got: `You must set leaderElection to use more than 1 replica`
```yaml
replicas: 2
leaderElection:
  enabled: false
```

4. Works smoothly

```yaml
replicas: 2
leaderElection:
  enabled: true
``

5. Works. We allowed this case since user might want to scale their deployment later
```
replicas: 1
leaderElection:
  enabled: true
```

Comment (eminaktas) [reply charts/descheduler/templates/deployment.yaml]: We've also added a note that we think may be useful to users.

```
{{- if eq .Values.kind "Deployment" }}
{{- if eq .Values.replicas 1.0}}
WARNING: You set replica count as 1 and workload kind as Deployment however leaderElection is not enabled. Consider enabling Leader Election for HA mode.
{{- end}}
{{- end}}
```

File: pkg/descheduler/descheduler.go [hunk 7]
```diff
@@ -69,7 +69,22 @@ func Run(ctx context.Context, rs *options.DeschedulerServer) error {
...
-	return RunDeschedulerStrategies(ctx, rs, deschedulerPolicy, evictionPolicyGroupVersion)
+	runFn := func() error {
+		return RunDeschedulerStrategies(ctx, rs, deschedulerPolicy, evictionPolicyGroupVersion)
+	}
+
+	if rs.LeaderElection.LeaderElect && rs.DeschedulingInterval.Seconds() == 0 {
+		return fmt.Errorf("leaderElection must be used with deschedulingInterval")
+	}
+
+	if rs.LeaderElection.LeaderElect {
```
Comment (ingvagabund) [inline pkg/descheduler/descheduler.go:80]: `if rs.LeaderElection.LeaderElect && !rs.DryRun {`

No need to have the leader election on when the descheduler is run in the dry run mode. There's a use case for combination of the dry-run mode and the metrics enabled when two or more deschedulers (each with a different configuration) are run in parallel with metrics reported with goal to find a configuration which leads to an optional eviction.

File: pkg/descheduler/leaderelection.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	LeaderElectionConfig *componentbaseconfig.LeaderElectionConfiguration,
+	ctx context.Context,
+) error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return fmt.Errorf("unable to get hostname: %v", err)
+	}
+
+	id := fmt.Sprintf("%s-%s", hostname, uuid.New().String())
+
```
Comment (ingvagabund) [inline pkg/descheduler/leaderelection.go:44]:
```go
if len(LeaderElectionConfig.Namespace) == 0 {
	return fmt.Errorf("namespace may not be empty")
}
if len(LeaderElectionConfig.Name) == 0 {
	return fmt.Errorf("name may not be empty")
}
```

File: pkg/descheduler/leaderelection.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+	client clientset.Interface,
+	LeaderElectionConfig *componentbaseconfig.LeaderElectionConfiguration,
+	ctx context.Context,
+) error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return fmt.Errorf("unable to get hostname: %v", err)
+	}
+
+	id := fmt.Sprintf("%s-%s", hostname, uuid.New().String())
```
Comment (ingvagabund) [inline pkg/descheduler/leaderelection.go:43]:
```go
import "k8s.io/apimachinery/pkg/util/uuid"
...
if hostname, err := os.Hostname(); err != nil {
	// on errors, make sure we're unique
	id = string(uuid.NewUUID())
} else {
	// add a uniquifier so that two processes on the same host don't accidentally both become active
	id = hostname + "_" + string(uuid.NewUUID())
}
```

**Review by ingvagabund — CHANGES_REQUESTED** [review]

**Comment by ingvagabund** [comment]
@Dentrax @eminaktas would you mind extending the [e2e test suite](https://github.com/kubernetes-sigs/descheduler/tree/master/test/e2e) with a test validating only a single instance is active at a time? E.g.
- create 2 namespaces (e.g. A, B)
- create e.g. 5 pods in each
- wait for 5 or more seconds, configure both deschedulers to evict pods running longer than 5 seconds, each descheduler targeting a specific namespace (includeNamespaces)
- validate only pods from a single namespace (A xor B) are evicted.

This will be a bit tricky as you will need to invoke `Run` from the `pkg/descheduler/descheduler.go` or some higher function so you are running both deschedulers from within the same binary.

**Comment by Dentrax** [comment]
> @Dentrax @eminaktas would you mind extending the [e2e test suite](https://github.com/kubernetes-sigs/descheduler/tree/master/test/e2e?rgh-link-date=2022-03-14T09%3A38%3A02Z) with a test validating only a single instance is active at a time?

We've tried to write a simple test by following your instructions, here is test output:

```
=== RUN   TestLeaderElection
    e2e_leaderelection_test.go:45: Creating testing namespace e2e-testleaderelection-a
    e2e_leaderelection_test.go:52: Creating testing namespace e2e-testleaderelection-b
    e2e_leaderelection_test.go:175: Creating deployment leaderelection for namespace e2e-testleaderelection-a
    e2e_leaderelection_test.go:175: Creating deployment leaderelection for namespace e2e-testleaderelection-b
    e2e_leaderelection_test.go:203: Pod leaderelection-76dbf88fff-kpv7d not running yet, is Pending instead
    e2e_leaderelection_test.go:203: Pod leaderelection-76dbf88fff-v98g4 not running yet, is Pending instead
    e2e_leaderelection_test.go:94: starting deschedulers
I0317 12:40:08.240823   75207 leaderelection.go:248] attempting to acquire leader lease kube-system/descheduler...
I0317 12:40:08.270859   75207 leaderelection.go:258] successfully acquired lease kube-system/descheduler
I0317 12:40:11.232282   75207 leaderelection.go:248] attempting to acquire leader lease kube-system/descheduler...
    e2e_leaderelection_test.go:135: Total evicted Pods count for e2e-testleaderelection-a: is 5 and e2e-testleaderelection-b: is 14
--- PASS: TestLeaderElection (43.24s)
PASS
ok      sigs.k8s.io/descheduler/test/e2e        44.618s
```

We also had to put two new policy files since we couldn't pass a `PolicyConfig` into `DeschedulerServer` struct, so we bring the following files:

```
test/e2e/policy_leaderelection_a.yaml
test/e2e/policy_leaderelection_b.yaml
```

Eventually, we're asserting that only pods from a single namespace are evicted:
```go
singleNamespaceEvicted := (podsA == 5 && podsB != 5) || (podsA != 5 && podsB == 5)
```


Thanks, @ingvagabund! I think we've resolved the following above reviews.

**Comment by a7i** [comment]
Re: RBAC, I would have expected the `Role` definition to include the following:
```
  - apiGroups: ["coordination.k8s.io"]
    resources: ["leases"]
    verbs: ["create"]
  - apiGroups: ["coordination.k8s.io"]
    resources: ["leases"]
    resourceNames: ["descheduler"]
    verbs: ["get", "patch", "delete"]
```

File: test/e2e/e2e_leaderelection_test.go [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+	s, err := options.NewDeschedulerServer()
+	if err != nil {
+		t.Fatalf("unable to initialize server: %v", err)
+	}
+	s.Client = clientSet
+	s.DeschedulingInterval = 5 * time.Second
+	s.LeaderElection.LeaderElect = true
+	s.KubeconfigFile = os.Getenv("KUBECONFIG")
+
+	s1 := s
```
Comment (ingvagabund) [inline test/e2e/e2e_leaderelection_test.go:90]:
```go
s1, err := options.NewDeschedulerServer()
s2, err := options.NewDeschedulerServer()
```
rather than copying `s` to avoid sharing data.

**Review by ingvagabund — CHANGES_REQUESTED** [review]

File: test/e2e/e2e_leaderelection_test.go (hunk 10 above)
Comment (Dentrax) [reply test/e2e/e2e_leaderelection_test.go:90]: Actually, I did it intentionally to avoid duplication for the following lines:

```go
if err != nil {
		t.Fatalf("unable to initialize server: %v", err)
	}
	s.Client = clientSet
	s.DeschedulingInterval = 5 * time.Second
	s.LeaderElection.LeaderElect = true
	s.KubeconfigFile = os.Getenv("KUBECONFIG")
```

Comment (ingvagabund) [reply test/e2e/e2e_leaderelection_test.go:90]: It is ok to duplicate those when it comes to the initialization. Other option is to create a function which will produce the same. Invoke twice then.

Comment (eminaktas) [reply test/e2e/e2e_leaderelection_test.go]: Hi @ingvagabund, I updated the changes as you mentioned.

File: pkg/descheduler/leaderelection.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+			},
+			OnStoppedLeading: func() {
+				klog.V(1).InfoS("Leader lost")
+			},
+			OnNewLeader: func(identity string) {
+				// Just got the lock
+				if identity == id {
+					return
+				}
+				klog.V(1).Infof("New leader elected: ", identity)
```
Comment (ingvagabund) [inline pkg/descheduler/leaderelection.go:94]: `"New leader elected: "` -> `"New leader elected: %v"` (missing %v)

File: cmd/descheduler/app/options/options.go [hunk 12]
```diff
@@ -58,7 +60,17 @@ func NewDeschedulerServer() (*DeschedulerServer, error) {
 }
 
 func newDefaultComponentConfig() (*componentconfig.DeschedulerConfiguration, error) {
-	versionedCfg := v1alpha1.DeschedulerConfiguration{}
+	versionedCfg := v1alpha1.DeschedulerConfiguration{
+		LeaderElection: componentbaseconfig.LeaderElectionConfiguration{
+			LeaderElect:       false,
+			LeaseDuration:     metav1.Duration{Duration: 15 * time.Second},
+			RenewDeadline:     metav1.Duration{Duration: 10 * time.Second},
+			RetryPeriod:       metav1.Duration{Duration: 2 * time.Second},
```
Comment (ingvagabund) [inline cmd/descheduler/app/options/options.go:68]: This will generate a lot of requests to the apiserver. What about to go with LeaseDuration=137s, RenewDealine=107s, RetryPeriod=26s by default? Those are the defaults we use in OpenShift by experience.

File: test/e2e/e2e_leaderelection_test.go [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+			t.Errorf("unable to start descheduler: %v", err)
+			return
+		}
+	}()
+
+	defer clientSet.CoordinationV1().Leases(s1.LeaderElection.ResourceNamespace).Delete(ctx, s1.LeaderElection.ResourceName, metav1.DeleteOptions{})
+	defer clientSet.CoordinationV1().Leases(s2.LeaderElection.ResourceNamespace).Delete(ctx, s2.LeaderElection.ResourceName, metav1.DeleteOptions{})
+
+	// wait for a while so all the pods are 5 seconds older
+	time.Sleep(5 * time.Second)
```
Comment (ingvagabund) [inline test/e2e/e2e_leaderelection_test.go:125]: 7 is better since 5s is exactly the boundary given the second descheduler instance is created after `time.Sleep(1 * time.Second)`.

**Comment by ingvagabund** [comment]
Few more nits I noticed. Durations is the most important here. We need to set those wisely to reduce the traffic between the descheduler instances and the apiserver.

File: cmd/descheduler/app/options/options.go (hunk 12 above)
Comment (Dentrax) [reply cmd/descheduler/app/options/options.go]: Thanks, updated. We forgot to recheck these values whether these appropriate or not. 

Did you use any calculation formula to get these values? For example, why `137` instead of `120` (which is 2m)?

Comment (ingvagabund) [reply cmd/descheduler/app/options/options.go]: https://github.com/openshift/enhancements/blob/master/CONVENTIONS.md#handling-kube-apiserver-disruption will help with the context though it is not a must. We can go lower though I would stick with at least 60s for the renew deadline and lease duration.
